src/OrleansRuntime/Catalog/ActivationData.cs

```diff
@@ -321,6 +322,8 @@ public GrainId Identity
 
         public ActivationAddress Address { get; private set; }
 
+        public IServiceProvider ServiceProvider => this.ServiceScope?.ServiceProvider;
```

Comment:
I believe it was https://github.com/dotnet/orleans/pull/2338 (and also https://github.com/dotnet/orleans/pull/1918 is related to this same philosophy).

Yup, a test kit could be useful but I would definitely not buy into that for this PR :)